File: apps/desktop/src/lib/tauri/system.ts
import type {
	AppInfo,
	AppPathsInfo,
	AppUpdateInfo,
	HealthCheck,
	SystemInfo,
} from "@/types/app";

import { invokeCommand } from "./client";

export function getAppInfo() {
	return invokeCommand<AppInfo>("get_app_info");
}

export function getSystemInfo() {
	return invokeCommand<SystemInfo>("get_system_info");
}

export function getAppPaths() {
	return invokeCommand<AppPathsInfo>("get_app_paths");
}

export function healthCheck() {
	return invokeCommand<HealthCheck>("health_check");
}

export function checkAppUpdate() {
	return invokeCommand<AppUpdateInfo>("check_app_update");
}

export function openExternalUrl(url: string) {
	return invokeCommand<void>("open_external_url", { url });
}

export function getReleasesPageUrl() {
	return invokeCommand<string>("get_releases_page_url");
}

File: apps/desktop/src-tauri/src/commands/system.rs
use tauri::State;

use crate::application::services::SystemService;
use crate::domain::{AppInfo, AppPathsInfo, BrowserStatus, HealthCheck, NetworkInfo, SystemInfo};
use crate::error::AppError;
use crate::infrastructure::network::lookup_public_network_info;
use crate::state::AppState;

#[tauri::command]
pub async fn get_app_info() -> Result<AppInfo, AppError> {
    Ok(SystemService::get_app_info())
}

#[tauri::command]
pub async fn get_system_info() -> Result<SystemInfo, AppError> {
    Ok(SystemService::get_system_info())
}

#[tauri::command]
pub async fn get_app_paths(state: State<'_, AppState>) -> Result<AppPathsInfo, AppError> {
    Ok(SystemService::get_app_paths(&state.paths))
}

#[tauri::command]
pub async fn health_check(state: State<'_, AppState>) -> Result<HealthCheck, AppError> {
    SystemService::health_check(&state).await
}

#[tauri::command]
pub async fn get_browser_status(state: State<'_, AppState>) -> Result<BrowserStatus, AppError> {
    state.browser_service.status(&state).await
}

#[tauri::command]
pub async fn set_browser_executable(
    state: State<'_, AppState>,
    path: String,
) -> Result<BrowserStatus, AppError> {
    state.browser_service.set_executable(&state, path).await
}

#[tauri::command]
pub async fn get_network_info() -> Result<NetworkInfo, AppError> {
    lookup_public_network_info().await
}
